python/tests/test_tree_stats.py

```diff
@@ -3519,47 +3631,84 @@ def naive_site_allele_frequency_spectrum(
     num_windows = len(windows) - 1
+    num_time_windows = len(time_windows) - 1
     out_dim = [1 + len(sample_set) for sample_set in sample_sets]
     out = np.zeros([num_windows] + out_dim)
+    out = np.zeros([num_windows] + [num_time_windows] + out_dim)
     for j in range(num_windows):
         begin = windows[j]
         end = windows[j + 1]
-        S = np.zeros(out_dim)
-        trees = [
-            next(ts.trees(tracked_samples=sample_set)) for sample_set in sample_sets
-        ]
-        t = trees[0]
-        while True:
-            tr_len = min(end, t.interval.right) - max(begin, t.interval.left)
-            if tr_len > 0:
-                for node in t.nodes():
-                    if 0 < t.num_samples(node) < ts.num_samples:
-                        x = [tree.num_tracked_samples(node) for tree in trees]
-                        # Note x must be a tuple for indexing to work
-                        if not polarised:
-                            x = fold(x, out_dim)
-                        S[tuple(x)] += t.branch_length(node) * tr_len
-
-            # Advance the trees
-            more = [tree.next() for tree in trees]
-            assert len(set(more)) == 1
-            if not more[0]:
-                break
-        if span_normalise:
-            S /= end - begin
-        out[j, :] = S
+        for k, upper_time in enumerate(time_windows[1:]):
+            S = np.zeros(out_dim)
+            if np.isfinite(upper_time):
+                decap_ts = ts.decapitate(upper_time)
+            else:
+                decap_ts = ts
+            assert np.all(list(ts.samples()) == list(decap_ts.samples()))
+            trees = [
+                next(decap_ts.trees(tracked_samples=sample_set))
+                for sample_set in sample_sets
+            ]
+            t = trees[0]
+            while True:
+                tr_len = min(end, t.interval.right) - max(begin, t.interval.left)
+                if tr_len > 0:
+                    for node in t.nodes():
+                        if 0 < t.num_samples(node) < decap_ts.num_samples:
+                            x = [tree.num_tracked_samples(node) for tree in trees]
+                            if not polarised:
+                                x = fold(x, out_dim)
+                            # Note x must be a tuple for indexing to work
+                            S[tuple(x)] += t.branch_length(node) * tr_len
+
+                # Advance the trees
+                more = [tree.next() for tree in trees]
+                assert len(set(more)) == 1
+                if not more[0]:
+                    break
+            if span_normalise:
+                S /= end - begin
+            out[j, k, :] = S
+
+    if drop_time_windows:
+        # beware: this assumes the first dimension is windows
+        assert out.shape[1] == 1
+        out = out[:, 0]
+    elif drop_windows:
+        # drop windows dim if only using time windows
+        out = out[0]
+        # assert out.shape[0] == 1
```

Comment:
This should still be okay, no?
```suggestion
        assert out.shape[0] == 1
        out = out[0]
```